sync/src/types.rs

```diff
@@ -250,94 +252,307 @@ pub struct Peers {
-        self.states.get(block)
+    pub fn inflight_state_by_block(&self, block: &BlockNumberAndHash) -> Option<&InflightState> {
+        self.inflight_states.get(block)
     }
 
-    pub fn prune(&mut self) {
+    pub fn compact_reconstruct(&mut self, peer: PeerIndex, hash: Byte32) -> bool {
+        let entry = self.compact_reconstruct_inflight.entry(hash).or_default();
+        if entry.len() >= 2 {
+            return false;
+        }
+
+        entry.insert(peer)
+    }
+
+    pub fn remove_compact(&mut self, peer: PeerIndex, hash: &Byte32) {
+        self.compact_reconstruct_inflight
+            .get_mut(&hash)
+            .map(|peers| peers.remove(&peer));
+    }
+
+    pub fn inflight_compact_by_block(&self, hash: &Byte32) -> Option<&HashSet<PeerIndex>> {
+        self.compact_reconstruct_inflight.get(hash)
+    }
+
+    pub fn mark_slow_block(&mut self, tip: BlockNumber) {
+        let now = faketime::unix_time_as_millis();
+        for key in self.inflight_states.keys() {
+            if key.number > tip + 1 {
+                break;
+            }
+            self.trace_number.entry(key.clone()).or_insert(now);
+        }
+    }
+
+    pub fn prune(&mut self, tip: BlockNumber) -> HashSet<PeerIndex> {
         let now = unix_time_as_millis();
         let prev_count = self.total_inflight_count();
-        let blocks = &mut self.blocks;
-        self.states.retain(|k, v| {
-            let outdate = (v.timestamp + BLOCK_DOWNLOAD_TIMEOUT) < now;
-            if outdate {
-                for peer in &v.peers {
-                    blocks.get_mut(peer).map(|set| set.remove(k));
+        let mut disconnect_list = HashSet::new();
+
+        let trace = &mut self.trace_number;
+        let download_schedulers = &mut self.download_schedulers;
+        let states = &mut self.inflight_states;
+
+        let mut remove_key = Vec::new();
+        // Since this is a btreemap, with the data already sorted,
+        // we don't have to worry about missing points, and we don't need to
+        // iterate through all the data each time, just check within tip + 20,
+        // with the checkpoint marking possible blocking points, it's enough
+        let end = tip + 20;
+        for (key, value) in states.iter() {
+            if key.number > end {
+                break;
+            }
+            if value.timestamp + BLOCK_DOWNLOAD_TIMEOUT < now {
+                download_schedulers
+                    .get_mut(&value.peer)
+                    .map(|set| set.hashes.remove(key));
+                if !trace.is_empty() {
+                    trace.remove(&key);
                 }
+                disconnect_list.insert(value.peer);
+                remove_key.push(key.clone());
+            }
+        }
+
+        for key in remove_key {
+            states.remove(&key);
+        }
+
+        download_schedulers.retain(|k, v| {
+            // task number zero means this peer's response is very slow
+            let zero = v.task_count != 0;
```

Comment:
The variable name `zero` is misleading

```suggestion
            let zero = v.task_count == 0;
            if zero {
                disconnect_list.insert(*k);
            }
            !zero
```

or:

```suggestion
            if v.task_count == 0 {
              disconnect_list.insert(*k);
              false
            } else {
              true
            }
```